src/electos/ballotmaker/data/extractor.py

```diff
@@ -0,0 +1,311 @@
+            choices.append({
+                "id": selection.model__id,
+                "choice": choice,
+            })
+        district = self._contest_election_district(contest)
+        full_text = _text_content(contest.full_text)
+        result = {
+            "id": contest.model__id,
+            "type": "ballot measure",
+            "title": contest.name,
+            "district": district,
+            "text": full_text,
+            "choices": choices,
+        }
+        return result
+
+
+    def _contests(self, ballot_style: BallotStyle):
+        """Extract contest subset needed for ballots."""
+        for contest in self._ballot_style_contests(ballot_style):
+            if isinstance(contest, CandidateContest):
+                entry = self._candidate_contest_of(contest)
+            elif isinstance(contest, BallotMeasureContest):
+                entry = self._ballot_measure_contest_of(contest)
+            else:
+                # Ignore other contest types
+                print(f"Skipping contest of type {contest.model__type}")
+            yield entry
+
+
+    def _election_ballot_styles(self, election: Election):
+        """Extract all ballot styles."""
+        for ballot_style in election.ballot_style:
+            data = {
+                "id": self._ballot_style_external_id(ballot_style),
+                "scopes": [
+                    _.model__id
+                    for _ in self._ballot_style_gp_units(ballot_style)
+                ],
+                "contests": [
+                    _
+                    for _ in self._contests(ballot_style)
+                ],
+            }
+            yield data
+
+
+    def _elections(self, election_report: ElectionReport):
+        """Extract all elections."""
+        # In most cases there isn't more than one 'Election' in a report, but the
+        # standard allows more than one, so handle them.
+        for election in election_report.election:
+            data = {
+                "name": _text_content(election.name),
+                "type": election.type.value,
+                "start_date": election.start_date.strftime("%Y-%m-%d"),
+                "end_date": election.end_date.strftime("%Y-%m-%d"),
+                "ballot_styles": [
+                    _ for _ in self._election_ballot_styles(election)
+                ],
+            }
+            yield data
+
+
+    def extract(self, data: Dict, index: ElementIndex = None) -> ElectionData:
+        """Extract election data.
+
+        This is the primary entry point for the extractor.
+
+        Parameters:
+            data: An EDF / election report dictionary.
+            index: An ElementIndex.
+                If empty (the default), create a new index from the election report.
+                Use this parameter only if there's already an existing index.
+
+        Returns:
+            Election data model for use in ballot rendering.
+        """
```

Comment:
I'm glad that's helpful!

The other longer docstrings are because some of these functions have edge cases. This one is detailed because it's the public interface.